File: src/checker.generics.test.ts
import { test } from "node:test";
import { expect } from "expect";

import { createChecker } from "./checker.ts";
import { typeToString } from "./checkerTypes.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { getIdentifierAtPosition } from "./nodeAtPosition.ts";
import { createProgram } from "./program.ts";
import { type Identifier, type VariableDeclarator } from "./types.ts";

function setup(text: string) {
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument("file:///T.java", text, 1);
  return { checker: createChecker(program), sf: program.getSourceFile("file:///T.java")! };
}

function sym(ctx: ReturnType<typeof setup>, needle: string, occ = 1) {
  let offset = -1;
  for (let i = 0; i < occ; i++) offset = ctx.sf.text.indexOf(needle, offset + 1);
  return ctx.checker.resolveName(getIdentifierAtPosition(ctx.sf, offset) as Identifier)!;
}

// Type of the initializer of the `var zz = ...;` in the given class body.
function zzType(body: string): string {
  const ctx = setup(`import java.util.*;\nclass C<E> { E val; ${body} }`);
  const declarator = sym(ctx, "zz").valueDeclaration as VariableDeclarator;
  return typeToString(ctx.checker.getTypeOfExpression(declarator.initializer!));
}

// Declared type of the `zz` variable itself (exercises var inference).
function zzVarType(body: string): string {
  const ctx = setup(`import java.util.*;\nclass C { ${body} }`);
  return typeToString(ctx.checker.getTypeOfSymbol(sym(ctx, "zz")));
}

// --- member-access substitution ----------------------------------------------------

test("List<String>.get(int) substitutes E to String", () => {
  expect(zzType("void m(List<String> xs) { var zz = xs.get(0); }")).toBe("String");
});

test("Map<String, Integer>.get substitutes V to Integer", () => {
  expect(zzType("void m(Map<String, Integer> mp) { var zz = mp.get(null); }")).toBe("Integer");
});

test("ArrayList<String>.get (member on the class itself) substitutes E", () => {
  expect(zzType("void m(ArrayList<String> xs) { var zz = xs.get(0); }")).toBe("String");
});

test("ArrayList<String>.iterator() substitutes E to Iterator<String>", () => {
  expect(zzType("void m(ArrayList<String> xs) { var zz = xs.iterator(); }")).toBe(
    "Iterator<String>",
  );
});

test("List<String>.iterator() threads E through Collection and Iterable", () => {
  expect(zzType("void m(List<String> xs) { var zz = xs.iterator(); }")).toBe("Iterator<String>");
});

test("chained: List<String>.iterator().next() yields String", () => {
  expect(zzType("void m(List<String> xs) { var zz = xs.iterator().next(); }")).toBe("String");
});

test("own type parameter field substitutes through the receiver", () => {
  expect(zzType("void m(C<String> c) { var zz = c.val; }")).toBe("String");
});

test("nested generic argument substitutes (List<List<String>>.get -> List<String>)", () => {
  expect(zzType("void m(List<List<String>> xs) { var zz = xs.get(0); }")).toBe("List<String>");
});

// --- var inference -----------------------------------------------------------------

test("var infers a String initializer", () => {
  expect(zzVarType('void m() { var zz = "hi"; }')).toBe("String");
});

test("var infers an int initializer", () => {
  expect(zzVarType("void m() { var zz = 1 + 2; }")).toBe("int");
});

test("var infers from a generic member call", () => {
  expect(zzVarType("void m(java.util.List<String> xs) { var zz = xs.get(0); }")).toBe("String");
});

test("enhanced-for var infers the element type of a List<String>", () => {
  const ctx = setup(
    "import java.util.*;\nclass C { void m(List<String> xs) { for (var item : xs) { use(item); } } }",
  );
  expect(typeToString(ctx.checker.getTypeOfSymbol(sym(ctx, "item", 1)))).toBe("String");
});

test("enhanced-for var infers the element type of an array", () => {
  const ctx = setup("class C { void m(String[] arr) { for (var item : arr) { use(item); } } }");
  expect(typeToString(ctx.checker.getTypeOfSymbol(sym(ctx, "item", 1)))).toBe("String");
});

// --- generic method inference ------------------------------------------------------

test("generic method infers T from a reference argument", () => {
  expect(zzVarType('<T> T id(T arg) {} void m() { var zz = id("s"); }')).toBe("String");
});

test("generic method infers T from two arguments of the same type", () => {
  expect(zzVarType('<T> T pick(T a, T b) {} void m() { var zz = pick("a", "b"); }')).toBe("String");
});

test("generic method infers T from a List<T> argument's element", () => {
  expect(
    zzVarType(
      "<T> T first(java.util.List<T> xs) {} void m(java.util.List<String> ss) { var zz = first(ss); }",
    ),
  ).toBe("String");
});

test("generic method with no inference info degrades to the type variable", () => {
  // T appears only in the return, not in any parameter: stays unbound (T).
  expect(zzVarType("<T> T make() {} void m() { var zz = make(); }")).toBe("T");
});

test("unrelated argument leaves an unconstrained variable unbound", () => {
  expect(zzVarType('<T> int len(T arg) {} void m() { var zz = len("s"); }')).toBe("int");
});

File: togo/internal/compiler/checker_generics_test.go
package compiler

import (
	"strings"
	"testing"
)

// Port of src/compiler/checker.generics.test.ts.

func genericsSetup(text string) *checkerCtx {
	program := NewProgram()
	LoadJdkStub(program)
	program.SetOpenDocument("file:///T.java", text, 1)
	return &checkerCtx{program: program, checker: NewChecker(program), uri: "file:///T.java"}
}

func (ctx *checkerCtx) sym(needle string, occ int) *Symbol {
	sf := ctx.program.GetSourceFile(ctx.uri)
	text := sf.AsSourceFile().Text
	offset := -1
	for i := 0; i < occ; i++ {
		offset = strings.Index(text[offset+1:], needle) + offset + 1
	}
	return ctx.checker.ResolveName(GetIdentifierAtPosition(sf, offset))
}

func zzType(t *testing.T, body string) string {
	t.Helper()
	ctx := genericsSetup("import java.util.*;\nclass C<E> { E val; " + body + " }")
	declarator := ctx.sym("zz", 1).ValueDeclaration
	return typeToString(ctx.checker.GetTypeOfExpression(declarator.AsVariableDeclarator().Initializer))
}

func zzVarType(t *testing.T, body string) string {
	t.Helper()
	ctx := genericsSetup("import java.util.*;\nclass C { " + body + " }")
	return typeToString(ctx.checker.GetTypeOfSymbol(ctx.sym("zz", 1)))
}

func TestMemberAccessSubstitution(t *testing.T) {
	cases := []struct{ body, want string }{
		{"void m(List<String> xs) { var zz = xs.get(0); }", "String"},
		{"void m(Map<String, Integer> mp) { var zz = mp.get(null); }", "Integer"},
		{"void m(ArrayList<String> xs) { var zz = xs.get(0); }", "String"},
		{"void m(ArrayList<String> xs) { var zz = xs.iterator(); }", "Iterator<String>"},
		{"void m(List<String> xs) { var zz = xs.iterator(); }", "Iterator<String>"},
		{"void m(List<String> xs) { var zz = xs.iterator().next(); }", "String"},
		{"void m(C<String> c) { var zz = c.val; }", "String"},
		{"void m(List<List<String>> xs) { var zz = xs.get(0); }", "List<String>"},
	}
	for _, tc := range cases {
		if got := zzType(t, tc.body); got != tc.want {
			t.Errorf("%q -> %q, want %q", tc.body, got, tc.want)
		}
	}
}

func TestVarInference(t *testing.T) {
	cases := []struct{ body, want string }{
		{"void m() { var zz = \"hi\"; }", "String"},
		{"void m() { var zz = 1 + 2; }", "int"},
		{"void m(java.util.List<String> xs) { var zz = xs.get(0); }", "String"},
	}
	for _, tc := range cases {
		if got := zzVarType(t, tc.body); got != tc.want {
			t.Errorf("%q -> %q, want %q", tc.body, got, tc.want)
		}
	}
}

func TestEnhancedForVarInference(t *testing.T) {
	ctx := genericsSetup("import java.util.*;\nclass C { void m(List<String> xs) { for (var item : xs) { use(item); } } }")
	if got := typeToString(ctx.checker.GetTypeOfSymbol(ctx.sym("item", 1))); got != "String" {
		t.Errorf("List element var = %q, want String", got)
	}
	ctx2 := genericsSetup("class C { void m(String[] arr) { for (var item : arr) { use(item); } } }")
	if got := typeToString(ctx2.checker.GetTypeOfSymbol(ctx2.sym("item", 1))); got != "String" {
		t.Errorf("array element var = %q, want String", got)
	}
}

func TestGenericMethodInference(t *testing.T) {
	cases := []struct{ body, want string }{
		{"<T> T id(T arg) {} void m() { var zz = id(\"s\"); }", "String"},
		{"<T> T pick(T a, T b) {} void m() { var zz = pick(\"a\", \"b\"); }", "String"},
		{"<T> T first(java.util.List<T> xs) {} void m(java.util.List<String> ss) { var zz = first(ss); }", "String"},
		{"<T> T make() {} void m() { var zz = make(); }", "T"},
		{"<T> int len(T arg) {} void m() { var zz = len(\"s\"); }", "int"},
	}
	for _, tc := range cases {
		if got := zzVarType(t, tc.body); got != tc.want {
			t.Errorf("%q -> %q, want %q", tc.body, got, tc.want)
		}
	}
}
